test/data/small-testiment-test.tsv

```diff
@@ -0,0 +1,4 @@
+"plethora"	"respite"	"curate"	"domains"
+"ben&jerrys"	"curate"	"super_spotter"	"#saudi"
+"$650"	"plethora"	"man/woman"	"cheescake"
+"domains"	"super_spotter"	"respite"	"#saudi"
```

Comment:
Can you point me to where this file is used? 

Are these meant to represent tokenized sentences? 

I'd recommend having a normal sentence-ish form of:
```
The man/woman ate a plethora of cheesecake for $650.
Curate domains of #saudi.
super_spotter likes Ben&Jerrys.
```
Then running the examples thru the text transform, which will lowercase (by default), remove diacritics (by default), remove punctuation (enable punct=-), and tokenize (enable tokens=+). The text transform will output  a column named `[YOURINPUT]_normalizedText`, which then gets fed into the WordEmbedding transform. 

This is the pattern users should be following with their text as lowercasing, removing punctuation/diacritics, and tokenizing will get the best results. The options the user needs to pick depends on the WordEmbedding model, but all of the ones we ship will do the best w/ these options.

Reply:
This is used for the test in OnnxTests (OnnxTests.cs line 300). The reason why they are in such a weird form, is because at the moment text transform does not supporting saving as onnx yet (that will be added later with the new CSharp op), so I wanted to add a test that did not rely on any transforms outside of the WordEmbeddingsTransform itself. 

---
In reply to: [226729318](https://github.com/dotnet/machinelearning/pull/1249#discussion_r226729318) [](ancestors = 226729318)

Reply:
Are we testing any out-of-vocabulary words?

Reply:
Do not have any explicit tests checked in, but that was something I checked. The behavior for out-of-vocab words is that the resulting ONNX model will behave slightly weirdly if it has out of vocab words when executing it with Lotus (or any other RT). The missing word will have a -1 returned for it in the label encoder, which will then during gather use the weights for the 2nd to last word in the word pool instead. Obviously this is not ideal, but I am not sure if a better solution exists. 

The input words do not affect the resulting ONNX model, so checking for it from the ML.NET side does not make sense as it would require a large amount of additional computation.

The alternative would be to make the resulting ONNX model behave differently if it encounters an out of vocab word, but the only way to make it meaningfully behave differently is to make it crash. This is both a) arguably worse (single missing word crashing the whole pipeline), and b) much harder to implement as the ONNX structure would need to be modified (likely also making it significantly more complex and slower as a result) in order to create this specific behavior in an edge case.

The only improvement I can think of would be to make it use some predetermined value in case of a missing word (rather than the 2nd to last word in the pool), but I do not believe there is a specific universal "missing" value that would make more sense (since the transform takes the min, mean, and max of each value, setting it to 0, max, or average would still always affect the results).

---
In reply to: [226751849](https://github.com/dotnet/machinelearning/pull/1249#discussion_r226751849) [](ancestors = 226751849)

Reply:
Ooh, never mind, I see what the issue is. This is not consistent with what ML.NET does for missing words. I will work on double checking this and fixing it if necessary. Thanks!

---
In reply to: [226764148](https://github.com/dotnet/machinelearning/pull/1249#discussion_r226764148) [](ancestors = 226764148,226751849)

Reply:
Unfortunately there is no filter operator in ONNX. This does sound like a common/normal case though, and I am still looking into this. Unfortunately, it looks like I will have to make a workaround that is likely to greatly increase the complexity of the model. The current new model I am drafting would involve mapping not found words to float.max_value for the mins, float.min_value for the maxes, and to 0 for sums with the average being performed manually by division (adding up all the element values, adding up number of found words, and dividing the two). Currently still working on figuring out exactly what this will look like.

---
In reply to: [226786526](https://github.com/dotnet/machinelearning/pull/1249#discussion_r226786526) [](ancestors = 226786526)

Reply:
Thanks for checking in fixes for the out-of-vocabulary words. For the OOV words, what kind of test did you run? 

I'm thinking of running a tokenized sentence through both the ONNX & non-ONNX model:
* Mixed OOV -- `{ "the", " man/woman", " ate", " a", " plethora", " of", " cheesecake", " for", "$650" }`
* Fully OOV -- `{ "abc", "def" }`
* Fully in vacab -- `{ "plethora", "respite", "curate", "domains" }`
* Empty -- `{ }`